text/287-cli-deprecation-warnings.md

```diff
@@ -0,0 +1,224 @@
+
+### Does this mean my code will stop working after this change?
+
+No! This feature only adds warnings to the CLI,
+it does not result in any code that was previously working to now fail.
+
+Note that if you want to migrate to `V2` of the CDK,
+which we strongly recommend,
+you will have to handle all the warnings that this feature emits,
+as all deprecated elements in `V1` will be removed from `V2`,
+and thus no longer available to be used by your code.
+
+If you want to make sure your code does not use any deprecated APIs,
+and thus is ready for migrating to CDK `V2`,
+you can set the `CDK_DEPRECATED_ERROR` environment variable,
+which will make any CDK command that invokes synthesis
+(`cdk synth`, `cdk deploy`, `cdk diff`, etc.)
+fail with an exception if any deprecated element is used.
+
+## Internal FAQ
+
+### Why are we doing this?
+
+We are doing this to help customers have a smoother migration from `V1` to `V2` of the CDK.
+
+### Why should we _not_ do this?
+
+I see three reasons why we might not want to implement this feature:
+
+1. Deprecated elements are already highlighted by virtually any editor/IDE.
+  In my experience, customers are diligent in moving away from deprecated APIs,
+  so this feature might not provide much value.
+2. We risk overloading customers with warnings,
+  training them to ignore our output if there's too much of it.
+  A warning must be read in order to be effective, after all.
+3. Depending on the exact solution chosen
+  (see discussion below),
+  we might not be able to distinguish from the user's code using a deprecated API,
+  and from our own libraries using deprecated APIs.
+  Which means a user might get warnings that they will be unable to get rid of
+  (because they're coming from code that they don't control).
+
+### What changes are required to enable this change?
+
+There are two high-level components of this change:
+
+#### 1. Changes to JSII
+
+We will add a feature to JSII that adds the warning code to the emitted JavaScript for deprecated elements.
+It will be off by default, and will have to be activated explicitly to take effect.
+We will also add two options to the JSII CLI that allow passing the names of the environment variables used for,
+respectively, silencing the warnings, and turning the warnings into errors.
+
+#### 2. Using the changed JSII
+
+Once we have modified JSII and released a new version of it,
+we will need to use it in the CDK.
+We will have to start compiling CDK with the new option turned on,
+and also set the two additional options mentioned above to be
+`CDK_DEPRECATED_IGNORE` and `CDK_DEPRECATED_ERROR`, respectively.
+
+We should also modify our CDK test infrastructure to run with the `CDK_DEPRECATED_ERROR` option turned on by default,
+unless a test is explicitly checking a deprecated API as a regression test,
+in which case we should add an API for turning that option off for those tests.
```

Comment:
> add an API for turning that option off for those tests

Not following this. Add an API where? can you expand?

Reply:
We should discuss this f2f. 

I'm working on a similar area and have encountered unforeseen issues, and has generally been harder than I thought it will be. It's because of the different ways we deprecated elements.

Maybe you'll have the same problems, maybe not. Worth discussing.

Reply:
I meant add the API to our testing libraries. Similarly like you added an API for tests using feature flags.

Sure, I'd love to discuss this F2F.